README.md
# ztimespan
[![ztimespan](http://img.shields.io/npm/v/ztimespan.svg)](https://www.npmjs.org/package/ztimespan) [![ztimespan](http://img.shields.io/npm/dm/ztimespan.svg)](https://www.npmjs.org/package/ztimespan)

Timespans in JS.

[![ztimespan](https://nodei.co/npm/ztimespan.png?downloads=true&downloadRank=true)](https://nodei.co/npm/ztimespan/)

## Installation
`npm install --save ztimespan`
(Can also be used in browser, just add it as `src` of a `<script>` and it will be added to `window` as `Time`)

## Usage
If on a browser, Time will be in the global namespace. If not, then you can `require` it in Node:
```js
var Time = require("ztimespan");
```
### Constructor
Time (or however you call the variable) will be a function, that can be called with `new` (such as `new Time()`) and can also be called normally (such as `Time()`). Both do the same thing. It takes one optional argument to be the initial value for the Time instance (if not specified, it's 0). It can be one of:
  * A number (in milliseconds);
  * A Date (Gets amount of time since Jan 1, 1970);
  * Another Time instance (Gets the time stored in it);
  * An array of time unit and amount of it (e.g.: `["years", 1]` for 1 year).

Here are all the valid time units, just for reference:
  * `years`, `year`, `y`;
  * `months`, `month`, `mo`;
  * `weeks`, `week`, `w`;
  * `days`, `day`, `d`;
  * `hours`, `hour`, `h`;
  * `minutes`, `minute`, `mins`, `min`, `m`;
  * `seconds`, `second`, `secs`, `sec`, `s`.

P.S.: Using a time unit that is in singular doesn't change anything.

Example:
```js
var Time = require("ztimespan");
var myTimeSpan = new Time(5); // or just Time(5)
```
Time will also have a few utility functions as properties:
* `years(amount)`, `months(amount)`, `weeks(amount)`, `days(amount)`, `hours(amount)`, `minutes(amount)`, `seconds(amount)` -> Returns `amount` of said unit, in milliseconds.
* `validUnit(str)` -> Returns `true` if `str` is a valid time unit string, `false` otherwise. E.g.: `Time.validUnit("min")` returns `true`.

Example:
```js
Time.years(2);
Time.months(4);
Time.weeks(5);
Time.days(6);
Time.hours(9);
Time.minutes(1);
Time.seconds(7);
```

Time instances have the `time` property, which is the time stored, in milliseconds.
NOTE: Make sure to not make `time` negative. Most methods check if time is negative and set it to 0 if so, but keeping it negative can cause buggy behavior. (Same goes for NaN.)

### Methods

Now, let's go in detail on the Time class. It has the following methods:
  * add;
  * remove;
  * subtract;
  * clear.

P.S.: They are all chainable (as in you can do `.add(...).remove(...).subtract(...).clear(...)`).
P.P.S.: `Subtract` is the same as `remove`.

Both `add` and `remove`/`subtract` work the same way, except one is the opposite of the other: `add` is to add to stored time while `remove`/`subtract` is to remove. There are two ways you can call them:
1. `(timeUnit, amount)` -> Add/remove a certain amount of a time unit.
Example:
```js
myTimeSpan.add("year", 3); // add 3 years
myTimeSpan.remove("y", 2); // remove/subtract 2 years
myTimeSpan.subtract("years", 6); // remove/subtract 6 years
```

2. `(quantity)` -> A certain amount of time to add/remove. That can be:
* A Date (Gets amount of time since Jan 1, 1970);
* A Time instance (that can also be itself, gets time stored);
* A number (in milliseconds).
Example:
```js
myTimeSpan.add(new Date());
myTimeSpan.add(Time(1000));
myTimeSpan.add(1000);
```

The `clear` method doesn't take any argument and sets the time stored to 0 (I mean, you can always do `<Time object>.time = 0`, but this has the advantage of being chainable).

### Getter Properties

Time instances have the following properties (they are all read-only because they are getters):
* `units` -> Returns an object that indicates the amount of each unit.
For example:
```js
{
  years: 0,
  months: 0,
  weeks: 5,
  days: 1,
  hours: 0,
  minutes: 59,
  seconds: 3
}
```
* `years`, `months`, `weeks`, `days`, `hours`, `minutes`, `seconds` -> The amount of said unit that perfectly fits in the time stored.
* `totalYears`, `totalMonths`, `totalWeeks`, `totalDays`, `totalHours`, `totalMinutes`, `totalSeconds` -> The amount of said unit that fits in the time stored, perfectly or not. Basically dividing the time stored by 1 of that unit.
* `date` -> The Date representation of this timespan.

### String-ified

The Time class has the `toString` function returning a human-readable string of the timespan. Consider:
```js
var Time = require("ztimespan");
var myOtherTimeSpan = new Time(["y", 4, "m", 7]);
console.log(myOtherTimeSpan.toString());
```
This will output in console:
```
4 years and 7 minutes
```

### Natural operations

You can do natural operations (`+`, `-`, `*`, `/`, etc) with time instances, such as `time * 5`. However, keep in mind that it does those operations to the millisecond value AND returns the new _millisecond_ value, which means it doesn't return a time instance. However, you can always do:
```js
otherTime = new Time(time * 5);
```

### v2.0.0

This version has a very breaking change: It changes the value of a month to be 30 days instead of 28, which was pretty wrong (not all months are february :P), which changes the millisecond value for everything. To keep using it in case you already used this module before and don't want to break saved values, you can use the `Time#toggleCompat` function or `Time#setCompat` (which asks for a boolean argument and sets the use of compatibility to it). When compatibility is on _(NOTE: Compatibility goes through everywhere that it is loaded!)_, it changes to operate with the old time values. Please note that the instances are not changed, just the methods. And yes, you can turn off compatibility with the same way you can turn it on.


CHANGES.md
# Changelog

Just to organize things.

# 2.0
## 2.1.0
#### Initial (2.1.0)
Added normal operations (`+`, `-`, `*`, `/`, etc) to Time instances.

## 2.0.0
#### Initial (2.0.0)
Changed the amount of time that one month takes from 4 weeks to 30 days (which is around 4.2 weeks). To use the old value, use the `Time#toggleCompat` or `Time#setCompat` methods. The first one toggles it and the second one takes one boolean argument setting the use of old value to it.

# 1.0
## 1.2.0
#### 1.2.1
Made `npm test` actually work...

#### Initial (1.2.0)
Added a `Time#subtract` as alias of `Time#remove`. Also, added `jest` tests and fixed some bugs that made the unit getters not work properly. (+ Added badges to README)

## 1.1.0
#### 1.1.1
Added a `Time#validUnit` method just in case you want to make sure a string is a valid time unit. Also added `mins` and `secs` as valid time units.

#### Initial (1.1.0)
Made typings export a Time property instead for the class, edited code to add Time.Time (a circular) and made `CHANGES.md`.

## 1.0.0
#### 1.0.7
A failed attempt of exporting the class (at typings).

#### 1.0.6
Fixed typings property at `package.json`.

#### 1.0.5
Added `NaN` checks.

#### 1.0.4
Added `MIT` license.

#### 1.0.3
A minor `README.md` list positioning fix.

#### 1.0.2
A minor `README.md` change (added inline code blocks to time units).

#### 1.0.1
Made `README.md`, added the `date` getter and added checks to see if a Time instance's `time` property is 0.

#### Initial (1.0.0)
The package is made.


test/ztimespan.test.js
var Time = require("../src/index");

describe("Constructor", function() {
  test("with 'new'", function() {
    var newTime = new Time();
    expect(newTime).toBeTruthy();
    expect(newTime).toBeInstanceOf(Time);
    expect(newTime.time).toBe(0);
  });
  test("without 'new'", function() {
    var time = Time();
    expect(time).toBeTruthy();
    expect(time).toBeInstanceOf(Time);
    expect(time.time).toBe(0);
  });
  test("equal with and without 'new'", function() {
    var newTime = new Time();
    var time = Time();
    expect(newTime.time).toBe(time.time);
    expect(newTime).toMatchObject(time);
  });
});

describe("Methods", function() {
  describe("Add", function() {
    var time = Time(0);
    test("adds 2 to 0 to become 2", function() { // Adding to 0
      time.add(2);
      expect(time.time).toBe(2);
    });
    test("adds 5 to 2 to become 7", function() { // Adding to existing value
      time.add(5);
      expect(time.time).toBe(7);
    });
    test("adds -8 to 7 to become 0 (reset)", function() { // Resetting cuz negative
      time.add(-8);
      expect(time.time).toBe(0);
    });
  });
  describe("Remove/Subtract", function() {
    test("remove has same behavior as subtract", function() {
      var time1 = Time(10);
      var time2 = Time(10);
      time1.remove(5);
      time2.subtract(5);
      expect(time1.time).toBe(time2.time);
    });
    var time = Time(10);
    test("removes 2 from 10 to become 8", function() { // Removing from existing value
      time.remove(2);
      expect(time.time).toBe(8);
    });
    test("removes 9 from 8 to become 0 (reset)", function() { // Resetting cuz negative
      time.remove(9);
      expect(time.time).toBe(0);
    });
  });
  describe("Clear", function() {
    var time = new Time(14);
    test("clears correctly", function() {
      time.clear();
      expect(time.time).toBe(0);
    });
  });
});
describe("NaN Resetting", function() {
  var time = new Time(1);
  time.time = NaN;
  test("resets on add", function() {
    time.add(0);
    expect(time.time).toBe(0);
  });
  time.time = "B";
  test("resets on remove/subtract", function() {
    time.remove(0);
    expect(time.time).toBe(0);
  });
  describe("'Units' getter", function() {
    time.time = NaN;
    test("resets on NaN", function() {
      time.units;
      expect(time.time).toBe(0);
    });
    time.time = -1;
    test("resets on negative", function() {
      time.units;
      expect(time.time).toBe(0);
    });
  });
  describe("toString", function() {
    time.time = NaN;
    test("resets on NaN", function() {
      time.toString();
      expect(time.time).toBe(0);
    });
    time.time = -1;
    test("resets on negative", function() {
      time.toString();
      expect(time.time).toBe(0);
    });
  });
  // obviously not testing on clear because that's already it's job to set to 0
});
describe("Getters", function() {
  var time = new Time(["y", 4, "mo", 2, "w", 3, "d", 6, "h", 12, "m", 30, "s", 2]);
  test("units", function() {
    expect(time.units).toMatchObject({
      years: 4,
      months: 2,
      weeks: 3,
      days: 6,
      hours: 12,
      minutes: 30,
      seconds: 2
    });
  });
  test("each unit", function() {
    expect(time.years).toBe(4);
    expect(time.months).toBe(2);
    expect(time.weeks).toBe(3);
    expect(time.days).toBe(6);
    expect(time.hours).toBe(12);
    expect(time.minutes).toBe(30);
    expect(time.seconds).toBe(2);
  });
  test("each total unit", function() {
    expect(time.totalYears).toBeCloseTo(4.243113490226338);
    expect(time.totalMonths).toBeCloseTo(50.91736188271605);
    expect(time.totalWeeks).toBeCloseTo(218.2172652116402);
    expect(time.totalDays).toBeCloseTo(1527.5208564814816);
    expect(time.totalHours).toBeCloseTo(36660.500555555554);
    expect(time.totalMinutes).toBeCloseTo(2199630.033333333);
    expect(time.totalSeconds).toBeCloseTo(131977802);
  });
  test("date conversion has same timestamp", function() {
    expect(time.date.getTime()).toBe(time.time);
  });
});
describe("Others", function() {
  var time = new Time(["y", 4, "mo", 2, "w", 3, "d", 6, "h", 12, "m", 30, "s", 2]);
  test("toString", function() {
    expect(time.toString()).toBe("4 years, 2 months, 3 weeks, 6 days, 12 hours, 30 minutes and 2 seconds");
  });
});


typings.d.ts
export as namespace Time;

export declare class Time {
  constructor(initialVal?: Time.InitialTimeValue);
  // properties
  public time: number;
  // getters
  public readonly years: number;
  public readonly totalYears: number;
  public readonly months: number;
  public readonly totalMonths: number;
  public readonly weeks: number;
  public readonly totalWeeks: number;
  public readonly days: number;
  public readonly totalDays: number;
  public readonly hours: number;
  public readonly totalHours: number;
  public readonly minutes: number;
  public readonly totalMinutes: number;
  public readonly seconds: number;
  public readonly totalSeconds: number;
  public readonly date: Date;
  public readonly units: { [prop in Time.TimeUnit]?: number };
  // methods
  public add: Time.TimeAddable<this>;
  public remove: Time.TimeAddable<this>;
  public subtract: Time.TimeAddable<this>;
  public clear(): this;
  public valueOf(): number;
  public toString(): string;
  // static
  public static years: Time.UnitFunc;
  public static months: Time.UnitFunc;
  public static weeks: Time.UnitFunc;
  public static days: Time.UnitFunc;
  public static hours: Time.UnitFunc;
  public static minutes: Time.UnitFunc;
  public static seconds: Time.UnitFunc;
  public static validUnit(str: string): str is Time.TimeUnit;
}

declare namespace Time {
  export type TimeUnit =
    "year"      | "years"  | "y"
    | "months"  | "month"  | "mo"
    | "weeks"   | "week"   | "w"
    | "days"    | "day"    | "d"
    | "hours"   | "hour"   | "h"
    | "minutes" | "minute" | "m" | "mins" | "min"
    | "seconds" | "second" | "s" | "secs" | "sec";
  export type InitialTimeValue = Time
    | Date
    | number
    | [TimeUnit, number];
  export type TimeAddable<T> = {
    (unit: TimeUnit, amount: number): T;
    (quantity: Time | Date | number): T;
  };
  export type UnitFunc = (amount: number) => number;
  /* export type TimeFuncType = {
    (initialVal?: InitialTimeValue): Time;
    new(initialVal?: InitialTimeValue): Time;
    years: UnitFunc;
    months: UnitFunc;
    weeks: UnitFunc;
    days: UnitFunc;
    hours: UnitFunc;
    minutes: UnitFunc;
    seconds: UnitFunc;
  };
  var TimeFunc: TimeFuncType; */
}


src/index.js
(function(){
  // constants
  var ADD_ERROR = "Unit or quantity must be a Time instance, Date instance, \
string (of unit, if using this add a second argument of the amount of that unit to use) or a number (milliseconds)!";
  var METHODS = ["years", "months", "weeks", "days", "hours", "minutes", "seconds"];
  var ALIASES = {
    years: ["years", "year", "y"],
    year: ["years", "year", "y"],
    y: ["years", "year", "y"],

    months: ["months", "month", "mo"],
    month: ["months", "month", "mo"],
    mo: ["months", "month", "mo"],

    weeks: ["weeks", "week", "w"],
    week: ["weeks", "week", "w"],
    w: ["weeks", "week", "w"],

    days: ["days", "day", "d"],
    day: ["days", "day", "d"],
    d: ["days", "day", "d"],

    hours: ["hours", "hour", "h"],
    hour: ["hours", "hour", "h"],
    h: ["hours", "hour", "h"],

    minutes: ["minutes", "minute", "mins", "min", "m"],
    minute: ["minutes", "minute", "mins", "min", "m"],
    mins: ["minutes", "minute", "mins", "min", "m"],
    min: ["minutes", "minute", "mins", "min", "m"],
    m: ["minutes", "minute", "mins", "min", "m"],

    seconds: ["seconds", "second", "secs", "sec", "s"],
    second: ["seconds", "second", "secs", "sec", "s"],
    sec: ["seconds", "second", "secs", "sec", "s"],
    secs: ["seconds", "second", "secs", "sec", "s"],
    s: ["seconds", "second", "secs", "sec", "s"],
  };
  // compatibility option
  var useCompat = false;
  // map
  var MAP = {
    years: 31104e6,
    months: 2592e6,
    weeks: 6048e5,
    days: 864e5,
    hours: 36e5,
    minutes: 6e4,
    seconds: 1e3
  };
  var MAP_ALT = {
    years: 290304e5,
    months: 24192e5
  };
  var appropriateNum = function(method) {
    let num;
    if (useCompat && method in MAP_ALT) {
      num = MAP_ALT[method];
    } else {
      num = MAP[method];
    }
    return num || 1e3;
  };
  // constructor
  var Time = function(initialVal) {
    this.time = 0;
    if (initialVal == null) {
      initialVal = 0;
    }
    if (this instanceof Time && !this.hasOwnProperty("__constructedTime")) {
      Object.defineProperty(this, "__constructedTime", {
        value: true,
        writable: false,
        configurable: false,
        enumerable: false
      });
      if (initialVal instanceof Date) {
        this.time = initialVal.getTime();
      } else if (initialVal instanceof Time) {
        this.time = initialVal.time;
      } else if (initialVal instanceof Array) {
        if (initialVal.length > 0) {
          for (var i = 0; i < initialVal.length; i += 2) {
            var item;
            try {
              item = initialVal[i].toString();
            } catch (err) {
              item = String(initialVal[i]);
            }
            if (!(item.toLowerCase() in ALIASES)) throw new TypeError('"' + item + '" is not a valid time unit!');
            if (i + 1 > initialVal.length) throw new Error('Amount for time unit "' + item + '" was not given!');
            var amount = initialVal[i + 1];
            if (isNaN(amount)) throw new TypeError('Amount for time unit "' + item + '" is not a number!');
            this.time += Time[ALIASES[item.toLowerCase()][0]](Number(amount));
          }
        }
      } else if (typeof initialVal === "number") {
        this.time = initialVal;
      } else {
        throw new TypeError("Initial value must be either a Date, a Time object, a number (in milliseconds)\
or an array of time unit and their amount.");
      }
    } else {
      return new Time(initialVal);
    }
  };
  Time.validUnit = function(str) {
    if (typeof str !== "string") return false;
    return (str.toLowerCase()) in ALIASES;
  };
  Time.setCompat = function(use) {
    const oldCompat = useCompat;
    useCompat = Boolean(use);
    return oldCompat;
  }
  Time.toggleCompat = function() {
    useCompat = !useCompat;
    return !useCompat;
  }
  Time.Time = Time;
  // methods
  var TP = Time.prototype;
  TP.add = function(unitOrQuantity, amount) {
    if (amount == null) amount = 0;
    if (unitOrQuantity instanceof Time) {
      this.time += unitOrQuantity.time;
    } else if (unitOrQuantity instanceof Date) {
      this.time += unitOrQuantity.getTime();
    } else if (typeof unitOrQuantity === "string") {
      unitOrQuantity = unitOrQuantity.toLowerCase();
      if (!(unitOrQuantity in ALIASES)) throw new TypeError("Invalid unit!");
      if (isNaN(amount)) throw new TypeError("Amount must be a number!");
      amount = Number(amount);
      this.time += Time[ALIASES[unitOrQuantity][0]](amount);
    } else if (typeof unitOrQuantity === "number") {
      if (isNaN(unitOrQuantity)) throw new TypeError("Quantity must not be NaN!");
      this.time += unitOrQuantity;
    } else {
      throw new TypeError(ADD_ERROR);
    }
    if (this.time < 0 || isNaN(this.time)) this.time = 0;
    return this;
  };
  TP.remove = function(unitOrQuantity, amount) {
    if (amount == null) amount = 0;
    if (unitOrQuantity instanceof Time) {
      return this.add(-unitOrQuantity.time);
    } else if (unitOrQuantity instanceof Date) {
      return this.add(-unitOrQuantity.getTime());
    } else if (typeof unitOrQuantity === "string") {
      return this.add(unitOrQuantity, -amount);
    } else if (typeof unitOrQuantity === "number") {
      return this.add(-unitOrQuantity);
    } else {
      throw new TypeError(ADD_ERROR);
    }
  };
  TP.subtract = TP.remove; // alias
  TP.clear = function() {
    this.time = 0;
    return this;
  };
  TP.valueOf = function() {
    if (this.time < 0 || isNaN(this.time)) this.time = 0;
    return Number(this.time); // in case it's null or undefined
  }
  TP.toString = function() {
    if (this.time < 0 || isNaN(this.time)) this.time = 0;
    var unitObj = this.units;
    var result = "";
    for (var i = 0; i < Object.keys(unitObj).length; i++) {
      var unit = Object.keys(unitObj)[i];
      if (!unitObj.hasOwnProperty(unit)) continue;
      var amount = unitObj[unit];
      if (amount === 0) {
        if (!result && unit === "seconds") {
          result = "0 seconds";
        }
        continue;
      }
      var strToUse = amount === 1 ? unit.replace(/s$/i, "") : unit;
      if (!result) {
        result = `${amount} ${strToUse}`;
      } else if (result.includes("and")) {
        result = result.replace(/\s*and(\s*\d+\s*\w+)$/i, ",$1 and");
        result += ` ${amount} ${strToUse}`;
      } else {
        result += ` and ${amount} ${strToUse}`;
      }
    }
    return result;
  };
  // getters
  var defineGetter = function(name, func) {
    Object.defineProperty(TP, name, { get: func });
  };
  defineGetter("units", function() {
    if (this.time < 0 || isNaN(this.time)) this.time = 0;
    var diff = this.time;
    var years = Math.floor(diff / (1000 * 60 * 60 * 24 * (useCompat ? 7 * 4 : 30) * 12));
    diff -= years * (1000 * 60 * 60 * 24 * (useCompat ? 7 * 4 : 30) * 12);

    var months = Math.floor(diff / (1000 * 60 * 60 * 24 * (useCompat ? 7 * 4 : 30)));
    diff -= months * (1000 * 60 * 60 * 24 * (useCompat ? 7 * 4 : 30));

    var weeks = Math.floor(diff / (1000 * 60 * 60 * 24 * 7));
    diff -= weeks * (1000 * 60 * 60 * 24 * 7);

    var days = Math.floor(diff / (1000 * 60 * 60 * 24));
    diff -= days * (1000 * 60 * 60 * 24);

    var hours = Math.floor(diff / (1000 * 60 * 60));
    diff -= hours * (1000 * 60 * 60);

    var mins = Math.floor(diff / (1000 * 60));
    diff -= mins * (1000 * 60);

    var seconds = Math.floor(diff / (1000));
    diff -= seconds * (1000);
    return {
      years: years,
      months: months,
      weeks: weeks,
      days: days,
      hours: hours,
      minutes: mins,
      seconds: seconds };
  });
  defineGetter("date", function() {
    return new Date(this.time);
  });
  for (var ind = 0; ind < METHODS.length; ind++) {
    var method = METHODS[ind];
    (function() { // sigh
      var actualMethod = method;
      Time[actualMethod] = function(amount) {
        return amount * appropriateNum(actualMethod);
      };
      defineGetter(actualMethod, function() {
        return this.units[actualMethod];
      });
      defineGetter("total" + actualMethod.replace(/^(\w)/, function(l) { return l.toUpperCase(); }), function() {
        return this.time / appropriateNum(actualMethod);
      });
    })();
  }
  if (typeof module === "object" && module.exports) {
    module.exports = Time;
  } else if (typeof window !== "undefined") {
    window.Time = Time;
  }
})();
